### ISSUE TITLE
Support `bfloat16` without mixed precision

### ISSUE THREAD

**Issue by pranavladkat** [tensorflow/tensorflow#58138]
### Describe the problem
I'm looking to train model with `bfloat16` datatype without having to use mixed precision on nvidia GPUs (A100 and beyond). This is important because directly using `bfloat16` datatype for all weights and intermediate tensors, I can reduce the memory by almost half compared to float32. I noticed I can set the float type using `tf.keras.backend.set_floatx()` API, however it currently does not support `bfloat16` datatype. 

Is this currently supported by tensorflow? If not, can this issue be considered as a feature request for the same?

Thanks

**Comment by pranavladkat** [comment]
After looking further into this, it seems the Adam optimizer is not working with `bfloat16` dtype, however SGD optimizer seems to be working. Please find an example script to reproduce this issue. 

Some config:
 - Using Tensorflow 2.10 branch
 - Compiled from source with cuda-11.2 and cudnn-8.4. 
 - AmazonLinux
 - Tested on Nvidia A100 GPUs
 
 I was not able to use tf-nightly to verify this since precompiled tf binaries are built with cudnn-8.1 whereas newer cudnn version is required for `bfloat16` support for conv2d layers. And I was not able to build latest master branch from source due to various build errors.
 
 
 ```
 # this example model converges after switching to SGD optimizer
 
import argparse
import tensorflow as tf
import sys
import time
import numpy as np
import math

print(tf.__file__)

bfloat16_t = tf.bfloat16.as_numpy_dtype


def set_model_weights(model, use_bfloat16):
    model_weights = model.get_weights()
    new_weights = []
    dtype = bfloat16_t if use_bfloat16 else 'float32'
    for weight in model_weights:
        w = np.random.normal(scale=1.0 / math.sqrt(float(weight.shape[0])), size=weight.shape).astype(dtype)
        new_weights.append(w)
    model.set_weights(new_weights)
    return model


@tf.function(jit_compile=True)
def training_step(model, loss, opt, images, labels):
    with tf.GradientTape() as tape:
        probs = model(images, training=True)
        loss_value = loss(labels, probs)
    gradients = tape.gradient(loss_value, model.trainable_variables)
    opt.apply_gradients(zip(gradients, model.trainable_variables))
    return loss_value, gradients


def run_eval(model, test_dataset):
    num_correct = 0
    for (images, labels) in test_dataset:
        num_correct += eval_step(model, images, labels)
    return num_correct


@tf.function(jit_compile=True)
def eval_step(model, images, labels):
    logits = model(images, training=False)
    correct = tf.equal(tf.argmax(logits, 1), labels)
    return tf.reduce_sum(tf.cast(correct, tf.int32))


def main(_):
    print(f"Tensorflow version : {tf.__version__}")
    tf.random.set_seed(args.seed)
    np.random.seed(args.seed)

    float_type = tf.float32
    if args.use_bfloat16:
        # tf.keras.mixed_precision.set_global_policy('mixed_bfloat16')
        tf.keras.mixed_precision.set_global_policy('bfloat16')
        float_type = tf.bfloat16

    # get rank and size of current process
    rank = 0
    size = 1

    gpus = tf.config.experimental.list_physical_devices('GPU')
    if gpus:
        device_id = 0
        tf.config.experimental.set_visible_devices(gpus[device_id], 'GPU')
        tf.config.experimental.set_memory_growth(gpus[device_id], True)

    (mnist_images, mnist_labels), (test_images, test_labels) = \
        tf.keras.datasets.mnist.load_data(path='mnist-%d.npz' % rank)
    dataset = tf.data.Dataset.from_tensor_slices(
        (tf.cast(mnist_images[..., tf.newaxis] / 255.0, float_type),
         tf.cast(mnist_labels, tf.int64))
    )
    dataset = dataset.shard(size, rank).shuffle(10000).batch(args.batch_size)

    test_dataset = tf.data.Dataset.from_tensor_slices(
        (tf.cast(test_images[..., tf.newaxis] / 255.0, float_type),
         tf.cast(test_labels, tf.int64))
    ).batch(args.batch_size)
    # test_dataset = dataset.shard(size, rank).shuffle(10000).batch(args.batch_size)

    mnist_model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, [3, 3], activation='relu'),
        tf.keras.layers.Conv2D(64, [3, 3], activation='relu'),
        tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
        tf.keras.layers.Dropout(0.25, seed=args.seed),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation='relu'),
        tf.keras.layers.Dropout(0.5, seed=args.seed),
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    loss = tf.losses.SparseCategoricalCrossentropy()
    opt = tf.optimizers.Adam(0.001 * size)
    ## model converges with SGD optimizer
    # opt = tf.optimizers.SGD(0.01)

    step = 0
    total_time_start = time.time()
    for epoch in range(1, int(args.num_epochs)+1):
        epoch_time_start = time.time()
        # train
        for (images, labels) in dataset:
            loss_value, grads = training_step(mnist_model, loss, opt, images, labels)
            # if step > 0:
            #     print(f"grad norms = {[tf.norm(g).numpy() for g in grads]}")
            loss_value = tf.cast(loss_value, tf.float32)
            if step == 0:
                for weight in mnist_model.trainable_variables:
                    print(f"weight name = {weight.name}, dtype = {weight.dtype}")
                mnist_model = set_model_weights(mnist_model, args.use_bfloat16)
                # broadcast variables from root to rest of the processes
            if step % 10 == 0 and rank == 0:
                print(f"Epoch {epoch} Step #{step} \tLoss: {loss_value:.6f}")
            step += 1
            # if step == 10:
            #     exit()
        epoch_time_end = time.time()
        # eval
        if rank == 0:
            correct_count = run_eval(mnist_model, test_dataset)
            print(f"Epoch {epoch} Eval accuracy {float(correct_count)/10000.0*100.0:.2f}% \t"
                  f"epoch time = {epoch_time_end - epoch_time_start:.3f}s")
    total_time_end = time.time()
    if rank == 0:
        print(f"Execution time: {(total_time_end - total_time_start):.3f}s")


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--batch-size', type=int, help='batch size used for training',
                        dest="batch_size", default=64)
    parser.add_argument('--num-epochs', type=int, help='Number of epochs', dest="num_epochs", default=1)
    parser.add_argument('--seed', type=int, help='random seed', dest="seed", default=17)
    parser.add_argument('--use-bfloat16', dest='use_bfloat16', action='store_true')
    parser.set_defaults(use_bfloat16=False)
    args, unparsed = parser.parse_known_args()
    print(f"args = {args}")
    main([sys.argv[0]] + unparsed)

 ```

**Comment by tilakrayal** [comment]
@pranavladkat,
The MatMul in the Dense layer silently runs on the CPU, since bfloat16 isn't supported on GPUs (except in XLA). You can verify this by calling tf.debugging.set_log_device_placement(True) at the start of the program, which prints the device for each op run.

Also please take a look at this [issue](https://github.com/tensorflow/tensorflow/issues/55738) and [comment](https://github.com/tensorflow/tensorflow/issues/56680#issuecomment-1233728566) from the developer with the similar feature for the refrence. Thank you

**Comment by pranavladkat** [comment]
Hi,

Thanks for the pointers. However, the provided links seem irrelevant to this issue. 

I have enabled XLA for training `tf.function(jit_compile=True)` and also I have used `tf.keras.mixed_precision.set_global_policy('bfloat16')` instead of `tf.keras.mixed_precision.set_global_policy('mixed_bfloat16')` to create weights with bfloat16 dtype. The reason I opted for this was setting `bfloat16` as float type through `tf.keras.backend.set_floatx()` is not yet supported. 

Also after verifying device placement through `tf.debugging.set_log_device_placement(True)` confirms the operations run on GPU.

**Comment by pdungen** [comment]
Hi,

I am facing the exact same issue. When using Adam optimizer and setting the model to bfloat16, the loss does not decrease. When set to SGD, it does decrease though.
I wanted to be double sure if its an architecture issue, so created the same network in Pytorch and it was able to decrease the loss.


```
from __future__ import print_function
import argparse
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.dropout1 = nn.Dropout(0.25)
        self.dropout2 = nn.Dropout(0.5)
        self.fc1 = nn.Linear(9216, 128)
        self.fc2 = nn.Linear(128, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = F.relu(x)
        x = F.max_pool2d(x, 2)
        x = self.dropout1(x)
        x = torch.flatten(x, 1)
        x = self.fc1(x)
        x = F.relu(x)
        x = self.dropout2(x)
        x = self.fc2(x)
        output = F.log_softmax(x, dim=1)
        return output


def train(args, model, device, train_loader, optimizer, epoch):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        if batch_idx % args.log_interval == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                epoch, batch_idx * len(data), len(train_loader.dataset),
                100. * batch_idx / len(train_loader), loss.item()))
            if args.dry_run:
                break


def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)

    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))


def main():
    # Training settings
    parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
    parser.add_argument('--batch-size', type=int, default=64, metavar='N',
                        help='input batch size for training (default: 64)')
    parser.add_argument('--test-batch-size', type=int, default=1000, metavar='N',
                        help='input batch size for testing (default: 1000)')
    parser.add_argument('--epochs', type=int, default=14, metavar='N',
                        help='number of epochs to train (default: 14)')
    parser.add_argument('--lr', type=float, default=1.0, metavar='LR',
                        help='learning rate (default: 1.0)')
    parser.add_argument('--gamma', type=float, default=0.7, metavar='M',
                        help='Learning rate step gamma (default: 0.7)')
    parser.add_argument('--no-cuda', action='store_true', default=False,
                        help='disables CUDA training')
    parser.add_argument('--no-mps', action='store_true', default=False,
                        help='disables macOS GPU training')
    parser.add_argument('--dry-run', action='store_true', default=False,
                        help='quickly check a single pass')
    parser.add_argument('--seed', type=int, default=1, metavar='S',
                        help='random seed (default: 1)')
    parser.add_argument('--log-interval', type=int, default=10, metavar='N',
                        help='how many batches to wait before logging training status')
    parser.add_argument('--save-model', action='store_true', default=False,
                        help='For Saving the current Model')
    args = parser.parse_args()
    use_cuda = not args.no_cuda and torch.cuda.is_available()
    use_mps = not args.no_mps and torch.backends.mps.is_available()

    torch.manual_seed(args.seed)

    if use_cuda:
        device = torch.device("cuda")
    elif use_mps:
        device = torch.device("mps")
    else:
        device = torch.device("cpu")

    train_kwargs = {'batch_size': args.batch_size}
    test_kwargs = {'batch_size': args.test_batch_size}
    if use_cuda:
        cuda_kwargs = {'num_workers': 1,
                       'pin_memory': True,
                       'shuffle': True}
        train_kwargs.update(cuda_kwargs)
        test_kwargs.update(cuda_kwargs)

    transform=transforms.Compose([
        transforms.ToTensor(),
        transforms.ConvertImageDtype(dtype= torch.bfloat16),
        ])
    dataset1 = datasets.MNIST('../data', train=True, download=True,
                       transform=transform)
    dataset2 = datasets.MNIST('../data', train=False,
                       transform=transform)
    train_loader = torch.utils.data.DataLoader(dataset1,**train_kwargs)
    # test_loader = torch.utils.data.DataLoader(dataset2, **test_kwargs)

    model = Net().to(dtype=torch.bfloat16, device=device)
    optimizer = optim.Adam(model.parameters(), lr=args.lr)

    for epoch in range(1, args.epochs + 1):
        train(args, model, device, train_loader, optimizer, epoch)
        # test(model, device, test_loader)


if __name__ == '__main__':
    main()
````

Not really sure what is wrong. Can also confirm with device placement that all the ops run on GPU.

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
